GetBonusActivation/handler.ts

```diff
@@ -0,0 +1,116 @@
+import {
+  withRequestMiddlewares,
+  wrapRequestHandler
+} from "io-functions-commons/dist/src/utils/request_middleware";
+import { readableReport } from "italia-ts-commons/lib/reporters";
+import {
+  IResponseErrorInternal,
+  IResponseErrorNotFound,
+  IResponseSuccessAccepted,
+  IResponseSuccessJson,
+  ResponseErrorInternal,
+  ResponseErrorNotFound,
+  ResponseSuccessAccepted,
+  ResponseSuccessJson
+} from "italia-ts-commons/lib/responses";
+import { FiscalCode } from "italia-ts-commons/lib/strings";
+import { BonusActivation } from "../generated/definitions/BonusActivation";
+import { BonusCode } from "../generated/definitions/BonusCode";
+import { BonusActivationModel } from "../models/bonus_activation";
+import { toApiBonusActivation } from "../utils/conversions";
+import { makeStartBonusActivationOrchestratorId } from "../utils/orchestrators";
+
+type IGetBonusActivationHandler = (
+  context: Context,
+  fiscalCode: FiscalCode,
+  bonusId: BonusCode
+) => Promise<
+  // tslint:disable-next-line: max-union-size
+  | IResponseSuccessJson<BonusActivation>
+  | IResponseErrorInternal
+  | IResponseErrorNotFound
+  | IResponseSuccessAccepted
+>;
+
+export function GetBonusActivationHandler(
+  bonusActivationModel: BonusActivationModel
+): IGetBonusActivationHandler {
+  return async (context, fiscalCode, bonusId) => {
+    const client = df.getClient(context);
+    const status = await client.getStatus(
+      makeStartBonusActivationOrchestratorId(fiscalCode)
+    );
+    return await tryCatch(
+      () =>
+        bonusActivationModel.findBonusActivationForUser(bonusId, fiscalCode),
+      err => new Error(`Query error: [${err}]`)
+    )
+      .chain(_ =>
+        fromEither(_).mapLeft(
+          queryError => new Error(`Query Error code=${queryError.code}`)
+        )
+      )
+      .fold<
+        // tslint:disable-next-line: max-union-size
+        | IResponseSuccessJson<BonusActivation>
+        | IResponseErrorInternal
+        | IResponseErrorNotFound
+        | IResponseSuccessAccepted
+      >(
+        err => {
+          context.log.error(`GetBonusActivation|ERROR|Error: [${err.message}]`);
+          return ResponseErrorInternal("Internal server error");
+        },
+        maybeBonusActivation => {
+          if (isSome(maybeBonusActivation)) {
+            return toApiBonusActivation(
+              maybeBonusActivation.value.bonusActivation
+            ).fold<
+              IResponseSuccessJson<BonusActivation> | IResponseErrorInternal
+            >(
+              err => {
+                context.log.error(
+                  `GetBonusActivation|ERROR|Conversion Error: [${readableReport(
+                    err
+                  )}]`
+                );
+                return ResponseErrorInternal("Error");
+              },
+              bonusActivation => ResponseSuccessJson(bonusActivation)
+            );
+          }
+          if (status.runtimeStatus === df.OrchestrationRuntimeStatus.Running) {
```

Comment:
can we move this check at the beginning (to avoid querying database in case) ?

Reply:
This check must be done after the query because it discriminates 202 response from 404 when that specific bonus doesn't exists. Otherwise, any new bonus activation for that user will block data retrieving of another existing bonus.

Reply:
can you add a comment with this statement?

Reply:
Fix in 0f2a09f